### DISCUSSION TITLE
How is the order of `test__` functions determined?

### DISCUSSION THREAD

**Discussion by jmhickman** [foundry-rs/foundry#3026] [Help] [answered]
I created a small testing contract (from @patrickalphac's foundry intro video) and was able to successfully run tests. However, as I added tests, I found that the order the `test*` functions is executed is not in the order they appear in the solidity test contract.

For example, in an effort to only test one assertion per function, I added a second test that checked that the address's balance of staked ERC20 token was in line with the amount that was deposited in the first test. However, this test always ran first, before the deposit had occurred. As a result, it always fails. 

Is there documentation about how `forge test` executes test functions, and how to force it to run certain tests in order?

**Comment by mds1** [comment, marked as answer]
Tests are executed in parallel, so you can't depend on order of test execution. (Note that individual runs of a fuzz test are run serially, however). You can use different contracts with different `setUp()` methods, or you can call `test1()` from within `test2()` to acquire its state, but this will result in test1 running twice.

Ideally your tests should be isolated and not dependent on each other, and having more than one `assertEq` per test is common 🙂

**Reply by jmhickman** [reply]
Thanks for the quick reply. 

Doesn't having multiple `asserts` in one function make it harder to figure out what actually went wrong? 

Related, is the state of the chain during tests reset in between functions? Or is the same chain 'maintained' for the duration of all tests and then dropped once execution is complete?

Thanks again for the reply, it was helpful.

**Reply by mds1** [reply]
> Doesn't having multiple asserts in one function make it harder to figure out what actually went wrong?

It can, but most assertions have an overload so you can add custom messages if you want like `assertEq(x, y, "balance mismatch")` https://github.com/dapphub/ds-test/blob/9310e879db8ba3ea6d5c6489a579118fd264a3f5/src/test.sol#L115

> Related, is the state of the chain during tests reset in between functions? Or is the same chain 'maintained' for the duration of all tests and then dropped once execution is complete?

State is reset between tests. The state after `setUp()` executes is the state used for all tests within a given contract

**Reply by mds1** [reply]
If you have a lot of assertions in a test and can't figure out what failed, another trick is to use `--watch` mode and gradually comment out assertions until you find the failing one(s)

**Reply by jmhickman** [reply]
This was really clarifying, thank you.
